habitat-hitl/habitat_hitl/core/ui_elements.py

```diff
@@ -0,0 +1,430 @@
+    for _ in range(user_count):
+        user_canvases.append(_create_default_canvases())
+    return user_canvases
+
+
+class UIManager:
+    """
+    Helper class for handling UI using UIElements.
+    """
+
+    def __init__(
+        self,
+        users: Users,
+        client_state: "RemoteClientState",
+        client_message_manager: "ClientMessageManager",
+    ):
+        self._client_state = client_state
+        self._client_message_manager = client_message_manager
+        self._users = users
+        self._user_canvases = _create_default_user_canvases(
+            users.max_user_count
+        )
+
+    def update_canvas(
+        self, canvas_uid: str, destination_mask: Mask
+    ) -> UIContext:
+        return UIContext(
+            canvas_uid=canvas_uid,
+            destination_mask=destination_mask,
+            manager=self,
+        )
+
+    def commit_canvas_content(
+        self,
+        canvas_uid: str,
+        ui_elements: Dict[str, UIElement],
+        destination_mask: Mask,
+    ):
+        for user_index in self._users.indices(destination_mask):
+            cached_elements = self._user_canvases[user_index].get(
+                canvas_uid, {}
+            )
+            self._user_canvases[user_index][canvas_uid] = ui_elements
+
+            def is_canvas_dirty() -> bool:
+                if len(cached_elements) != len(ui_elements):
+                    return True
+                else:
+                    return any(
+                        uid not in cached_elements
+                        or element != cached_elements[uid]
+                        for uid, element in ui_elements.items()
+                    )
+
+            if is_canvas_dirty():
+                dirty_elements: List[UIElementUpdate] = []
+
+                for element in ui_elements.values():
+                    dirty_elements.append(
+                        UIElementUpdate(
+                            canvasProperties=element
+                            if isinstance(element, UICanvas)
+                            else None,
+                            label=element
+                            if isinstance(element, UILabel)
+                            else None,
+                            toggle=element
+                            if isinstance(element, UIToggle)
+                            else None,
+                            button=element
+                            if isinstance(element, UIButton)
+                            else None,
+                            listItem=element
+                            if isinstance(element, UIListItem)
+                            else None,
+                            separator=element
+                            if isinstance(element, UISeparator)
+                            else None,
+                            spacer=element
+                            if isinstance(element, UISpacer)
+                            else None,
```

Comment:
This is not ideal.
See my comment about union serialization.

Reply:
Agreed, this is a bit ugly. Maybe a structure with default Nones and a map based override would be cleaner?